embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/logs/LogOrchestrator.kt

```diff
@@ -0,0 +1,72 @@
+package io.embrace.android.embracesdk.internal.logs
+
+import androidx.annotation.VisibleForTesting
+import io.embrace.android.embracesdk.internal.clock.Clock
+import io.embrace.android.embracesdk.worker.ScheduledWorker
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+
+internal class LogOrchestrator(
+    private val logOrchestratorScheduledWorker: ScheduledWorker,
+    private val clock: Clock,
+    private val sink: LogSinkImpl
+) {
+    private var lastLogTime: Long = 0
+    private var inactivityFuture: ScheduledFuture<*>? = null
+    private var batchTimeFuture: ScheduledFuture<*>? = null
+
+    init {
+        sink.callOnLogsStored(::onLogsAdded)
+    }
+    private fun onLogsAdded() {
+        lastLogTime = clock.now()
+        if (sink.completedLogs().size >= MAX_LOGS_PER_BATCH) {
+            sendLogs()
+        } else {
+            scheduleInactivityCheck()
+            if (batchTimeFuture == null) {
+                scheduleBatchTimeCheck()
+            }
+        }
+    }
+
+    @VisibleForTesting
+    fun sendLogs() {
```

Comment:
I'd suggest adding a `synchronized` block here on an `Any()` lock property. This would avoid logs getting sent multiple times if `scheduleInactivityCheck` and `scheduleBatchTimeCheck` are called at the same time

Reply:
Since `sink.flushLogs()` is `synchronized`, logs will be sent only once even if `sendLogs()` is called concurrently. I can add a comment here (and maybe a unit test) to ensure that it can not be accidentally refactored away. Or sync the whole method if we still think that's the best way to go.

Reply:
Awesome. A comment explaining that should be more than enough 👍